packages/nextjs/src/config/wrappers/wrapperUtils.ts

```diff
@@ -1,46 +1,58 @@
  */
-export async function wrapperCore<T extends DataFetchingFunction>(options: {
-  origFunction: T['fn'];
-  context: T['context'];
-  route: string;
-  op: string;
-}): Promise<T['result']> {
-  const { origFunction, context, route, op } = options;
+export function callDataFetcherTraced<F extends (...args: any[]) => Promise<any> | any>(
+  origFunction: F,
+  origFunctionArgs: Parameters<F>,
+  options: {
+    route: string;
+    op: string;
+  },
+): ReturnType<F> {
+  const { route, op } = options;
 
   const transaction = getActiveTransaction();
 
-  if (transaction) {
-    // TODO: Make sure that the given route matches the name of the active transaction (to prevent background data
-    // fetching from switching the name to a completely other route)
-    transaction.name = route;
-    transaction.metadata.source = 'route';
-
-    // Capture the route, since pre-loading, revalidation, etc might mean that this span may happen during another
-    // route's transaction
-    const span = transaction.startChild({ op, data: { route } });
-
-    // TODO: Can't figure out how to tell TS that the types are correlated - that a `GSPropsFunction` will only get passed
-    // `GSPropsContext` and never, say, `GSSPContext`. That's what wrapping everything in objects and using the generic
-    // and pulling the types from the generic rather than specifying them directly was supposed to do, but... no luck.
-    // eslint-disable-next-line prefer-const, @typescript-eslint/no-explicit-any
-    const props = await (origFunction as any)(context);
-
-    span.finish();
-
-    return props;
+  if (!transaction) {
+    return origFunction(...origFunctionArgs);
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  return (origFunction as any)(context);
+  // TODO: Make sure that the given route matches the name of the active transaction (to prevent background data
+  // fetching from switching the name to a completely other route) -- We'll probably switch to creating a transaction
+  // right here so making that check will probabably not even be necessary.
+  // Logic will be: If there is no active transaction, start one with correct name and source. If there is an active
+  // transaction, create a child span with correct name and source.
+  // We will probably need to put
+  transaction.name = route;
+  transaction.metadata.source = 'route';
+
+  // Capture the route, since pre-loading, revalidation, etc might mean that this span may happen during another
+  // route's transaction
+  const span = transaction.startChild({ op, data: { route } });
+
+  const result = origFunction(...origFunctionArgs);
+
+  // We do the following instead of `await`-ing the return value of `origFunction`, because that would require us to
+  // make this function async which might in turn create a mismatch of function signatures between the original
```

Comment:
This feels overly complicated to me. Whether the user provides a sync or an async function, all three of the data fetchers are set up to accept a promise:

`getInitialProps`: https://github.com/vercel/next.js/blob/de7aa2d6e486c40b8be95a1327639cbed75a8782/packages/next/shared/lib/utils.ts#L20

`getStaticProps`: https://github.com/vercel/next.js/blob/de7aa2d6e486c40b8be95a1327639cbed75a8782/packages/next/types/index.d.ts#L117-L123

`getServerSideProps`: https://github.com/vercel/next.js/blob/de7aa2d6e486c40b8be95a1327639cbed75a8782/packages/next/types/index.d.ts#L170-L176

So I think it's fine that we're always creating an async function. To handle the case of errors, we can just use a `try-catch-finally`. (It's a good thought - I'll update my PR to include it also.)

Reply:
I don't think this is overly complicated. If we have the opportunity not to change the function signature, why would we change it? It's not about what the Framework accepts, it's about not changing what the user provides. I don't really see the value behind adjusting this.

Reply:
Nextjs is going to have to await it, regardless, just in case it's async, though. 

To me, the question isn't why would we change it but why shouldn't we change it, if it makes no difference to nextjs and it allows us to us to simplify our code? Regardless of whether the alternative is "overly" complicated, it is certainly more complicated than simply doing `await original(...)`.

Having the simpler version is good for its own sake, but it will also make error handling simpler, because it'll allow us to just use `try-catch`, _and_ it brings us in line with what `withSentry` already does. (Eventually, I'd like to meld that into this mix also.)

Reply:
Alright. Changed in f45ceb6de3bf4fc429d507bfbd6f7e1667c72cf4